### DISCUSSION TITLE
setting pump speed, chlorinator value

### DISCUSSION THREAD

**Discussion by dwradcliffe** [jlvaillant/intellicenter#12] [Ideas]
The only thing missing now (for me) is setting the pump speed and setting the IntelliChlor output value.
I was thinking maybe we can add services to set those, but I just found there's a Number Entity which is probably more correct. Thoughts?

**Comment by dwradcliffe** [comment]
I was able to get a basic Number entity working to set the pump speed. Couple issues to solve still:

Setting the speed is a different object name and attribute than getting the current speed:
get = `PMP01`/`RPM`
set = `p0101`/`SPEED`

I'm not sure how to derive the second object name from config yet.
Once that's figured out it won't be hard to customize the logic for setting the value.

**Reply by jlvaillant** [reply]
I was never able to instantiate a Number entity (that was my first approach for setting the desired temperature as opposed to Heater). How did you do that?
The "work around" I used for doing things like "high speed pump mode" is to do it at the Pentair level and expose the setting as a feature circuit (which then appears as a switch in HA).
Could you send me the snippet of code with the NumberEntity thing you got?
And BTW, thanks for ironing out all the bugs on the IntelliChlor, I don't have that unit installed yet ;-)

**Reply by dwradcliffe** [reply]
WIP code is here https://github.com/jlvaillant/intellicenter/pull/14

**Reply by dwradcliffe** [reply]
Any opinions for how to read from / write to different objects?
I think we can see the write obj in the initial object list:

```json
{
      "objnam": "p0101",
      "params": {
        "OBJTYP": "PMPCIRC",
        "PARENT": "PMP01"
      }
    },
```

And the PARENT gives us the link to the read obj. Is it easy to access a different object from the entity object? Something like:
```python
    def value(self) -> float:
        pump_obj = self._controller.objects[self._poolObject['PARENT']]
        return pump_obj[RPM_ATTR]
```

**Reply by dwradcliffe** [reply]
After thinking on this, it's probably best to keep the entity linked with the object that it's reading from so that the push updates propagate properly. So I'll leave it as I have it in the PR. Still need to figure out a good way to detect the relationship though.

**Reply by jlvaillant** [reply]
AS far as I know, there is no 1 to 1 map between the pump and the setting of its speed. The 'p010X' subsection is I believe indexed by the "CIRCUIT" key, meaning that when the corresponding circuit is activated the speed setting in either GPM or RPM which be set for the pump. So, if you turn on the spa (C0001 for me), p0103 will activate, cleaner will be p0101 (again for my system), and so on... so changing the SPEED attribute for a 'p010X' will only affect the speed of the pump when the associated circuit is on. So I am not clear what you are trying to do, change the setting for the program or change the speed of the pump as it is running. For the later case (ie I want to have a different speed RIGHT NOW for my pump), the solution I used is to add a "virtual" circuit (call it "High Speed"), add a setting for the pump for that circuit and add the circuit as a feature. This way I have a Home Assistant "switch" I can turn on when I want a higher speed pump going.
I was trying to refrain into adding all the logic for configuring things like pump settings, schedules, body volumes and so on.
This is different I think for setting the Chlorinator value, which should be writable as a normal attribute on the object itself.
